### PR TITLE
refactor(storage): refactoring storage

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
**What type of PR is this?**

**Which issue does this PR fix**:
#Closes 

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/storage/storage_controller.go [hunk 1]
```diff
@@ -44,3 +52,46 @@ func (sc StoreController) GetImageStore(name string) ImageStore {
 
 	return sc.DefaultStore
 }
+
+// CheckIsImageSignature checks if the given image (repo:tag) represents a signature. The function
```
Comment (andaaron) [inline pkg/storage/storage_controller.go:56]: I'd say move this in storage.go, is unrelated to the storage controller.

File: pkg/storage/local/local.go [hunk 2]
```diff
@@ -34,31 +34,25 @@ import (
...
-	defaultSchemaVersion = 2
+	"zotregistry.io/zot/pkg/storage/types"
+	"zotregistry.io/zot/pkg/test/inject"
 )
 
 // ImageStoreLocal provides the image storage operations.
 type ImageStoreLocal struct {
-	rootDir     string
-	lock        *sync.RWMutex
-	blobUploads map[string]storage.BlobUpload
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/storage/local/local.go:53]: yes, I also wanted to remove it

File: pkg/api/routes_test.go [hunk 3]
```diff
@@ -22,7 +22,7 @@ import (
 	"zotregistry.io/zot/pkg/api/config"
 	"zotregistry.io/zot/pkg/api/constants"
 	localCtx "zotregistry.io/zot/pkg/requestcontext"
-	"zotregistry.io/zot/pkg/storage"
+	"zotregistry.io/zot/pkg/storage/types"
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/api/routes_test.go:25]: agree, you leave it as types in packages under storage/, but use storageTypes everywhere else

File: pkg/storage/common/common.go [hunk 4]
```diff
@@ -830,3 +784,33 @@ func (dt *dedupeTask) DoWork() error {
 
 	return err
 }
+
+func ValidateHardLink(rootDir string) error {
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/storage/common/common.go:788]: I'll move it under storage.go

File: pkg/storage/cache.go [hunk 5]
```diff
@@ -2,10 +2,57 @@ package storage
...
 import (
 	"zotregistry.io/zot/errors"
+	"zotregistry.io/zot/pkg/api/config"
 	zlog "zotregistry.io/zot/pkg/log"
 	"zotregistry.io/zot/pkg/storage/cache"
+	"zotregistry.io/zot/pkg/storage/constants"
 )
 
+func CreateCacheDatabaseDriver(storageConfig config.StorageConfig, log zlog.Logger) cache.Cache {
+	if storageConfig.Dedupe || storageConfig.StorageDriver != nil {
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/storage/cache.go:12]: I have kept the change that was the newest. In this case I saw that in main we have: 
```go
if storageConfig.StorageDriver != nil {
       params.Name = s3.CacheDBName
}
```
but in the newer commit from Sebi in his refactor PR we have this removed. 

Also I didn't think the cache should be called s3_cache if boltdb was used.

File: pkg/storage/storage.go [hunk 6]
```diff
@@ -1,59 +1,197 @@
...
+			// false positive lint - linter does not implement Lint method
+			//nolint: typecheck,contextcheck
+			defaultStore = s3.NewImageStore(rootDir, config.Storage.RootDirectory,
+				config.Storage.GC, config.Storage.GCDelay, config.Storage.Dedupe,
+				config.Storage.Commit, log, metrics, linter, store,
+				CreateCacheDatabaseDriver(config.Storage.StorageConfig, log))
+		}
+
+		storeController.DefaultStore = defaultStore
+	} else {
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/storage/storage.go:76]: I have refactored the check to return the error early

File: pkg/storage/cache.go [hunk 7]
```diff
@@ -2,10 +2,57 @@ package storage
...
+			params.Name = constants.BoltdbName
+			params.UseRelPaths = getUseRelPaths(&storageConfig)
+
+			driver, _ := Create("boltdb", params, log)
+
+			return driver
+		}
+
+		// remote cache
+		if storageConfig.CacheDriver != nil {
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/storage/cache.go:25]: I have removed the nested if blocks, I think we can keep the if statements to be more clear why we create the dynamo driver

[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/storage/cache.go]: ok

**Review by andaaron — DISMISSED** [review]
LGTM

**Review by peusebiu — APPROVED** [review]

**Review by peusebiu — APPROVED** [review]

File: pkg/api/controller_test.go [hunk 8]
```diff
@@ -51,8 +51,9 @@ import (
 	"zotregistry.io/zot/pkg/log"
 	"zotregistry.io/zot/pkg/meta/repodb/repodbfactory"
```
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/api/controller_test.go:52]: Refactoring pkg/meta is out of scope of this PR.

**Review by andaaron — APPROVED** [review]

File: pkg/api/controller_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/api/controller_test.go:52]: Laurentiu has a draft at: https://github.com/project-zot/zot/pull/1476

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**